### PR TITLE
DP4-2 [Admin] Create/Update product

### PR DESCRIPTION
**Pull request by koleshkos**
- Added button "Create"/"Edit"
- Validates for Product model:
  - Name
  - Image
  - Code
  - Price
  - Description
- Added Bootstrap

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -11,6 +12,36 @@
   </head>
 
   <body>
-    <%= yield %>
+    <header id="header" class="mb-3">
+      <nav class="navbar navbar-expand-lg navbar-dark bg-dark">
+        <div class="container">
+          <ul class="navbar-nav mr-auto">
+            <a class="navbar-brand" href="/">MℽSℍOℙ</a>
```
Comment (DmitryGolubko) [inline app/views/layouts/application.html.erb]: what a strange symbols in `MℽSℍOℙ`, use normal ones

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -11,6 +12,36 @@
...
+            <li><%= link_to 'Product', products_path, class: 'nav-link px-2 text-white'%></li>
```
Comment (DmitryGolubko) [inline app/views/layouts/application.html.erb]: Is it a link to a product list? then we should name it 'Products'

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -11,6 +12,36 @@
...
+      <%= yield %>
+    </div>
+
+    <footer class="fixed-bottom bg-dark">
+      <div class="container">
+        <div class="d-flex">
+          <div class="footer-left mt-2 mb-2">
+            <%= link_to 'FAQ', root_path, class: 'text-decoration-none text-white-50'%>
+            <%= link_to 'Help', root_path, class: 'text-decoration-none text-white-50'%>
+            <%= link_to 'About', root_path, class: 'text-decoration-none text-white-50'%>
```
Comment (DmitryGolubko) [inline app/views/layouts/application.html.erb]: Are we gonna have these pages? remove if not

File: app/views/products/new.html.haml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    = f.label :code
+    = f.text_field :code, class: 'form-control'
+    %br
+    = f.label :price
+    = f.number_field :price, class: 'form-control'
+    %br
+    = f.label :description
+    = f.text_field :description, class: 'form-control'
+    %br
+    = f.submit 'Save', class: 'btn btn-primary'
```
Comment (DmitryGolubko) [inline app/views/products/new.html.haml]: What's the difference between `new` and `edit` partials? Let's create separate partial 'views/products/_form' for form only and render it for new and edit actions.

File: app/views/products/show.html.haml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+= @product.name
+%br
+= image_tag(@product.image.url, size: "100x100")
+%br
+= @product.code
+%br
+= @product.price
```
Comment (DmitryGolubko) [inline app/views/products/show.html.haml]: Let's use `number_to_currency` helper there for better price view
https://apidock.com/rails/ActionView/Helpers/NumberHelper/number_to_currency

File: db/migrate/20220131094309_create_products.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class CreateProducts < ActiveRecord::Migration[6.1]
+  def change
+    create_table :products do |t|
+      t.string :name
+      t.string :image
+      t.string :code
+      t.integer :price
```
Comment (DmitryGolubko) [inline db/migrate/20220131094309_create_products.rb:9]: what if price will be with cents? let's use decimal here with `scale: 2`
https://nelsonfigueroa.dev/handling-decimal-precision-in-rails/

File: app/controllers/products_controller.rb [hunk 7]
```diff
@@ -1,5 +1,54 @@
 # frozen_string_literal: true
 
 class ProductsController < ApplicationController
-  def index; end
+  def index
+    @products = Product.all
+  end
+
+  def show
+    @product = Product.find(params[:id])
```
Comment (DmitryGolubko) [inline app/controllers/products_controller.rb]: let's move this finder to private method `get_product` which will be called for show, edit, update and destroy methods, use `before_filter` for this

**Review by DmitryGolubko — CHANGES_REQUESTED** [review]
Changes requested.

TODO: 
1. Remove all temp images from the github, github is not intended for storing images
2. Migrate image storing to AWS
3. Write specs for your model and controller, check that everything is working correct

**Comment by koleshkos** [comment]
- Removed all temp images from the github
- Migrated image storing to AWS
- Added before_action :set_product
- Used  decimal  with scale: 2 for price
- Used number_to_currency helper  for better price view
- Added partials form for new and edit
- Changed link name to 'Products'
- Removed unnecessary links in footer
- Fixed name project in header

Next step:
 - Writing specs for  model and controller

File: spec/controllers/products_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ProductsController, type: :controller do
+  let(:product) { Product.create(name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product') }
```
Comment (DmitryGolubko) [inline spec/controllers/products_controller_spec.rb]: move product creation to factories, use factory_bot and faker gems.
Do the same for products with bad data using traits

File: spec/controllers/products_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    let(:product) { Product.new }
+
+    it 'returns a success response' do
+      post :create, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product' } }
+      expect(response).to be_redirect
+    end
+
+    it 'with bad data' do
+      post :create, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: 1000, price: '12345RE', description: 'My Product' } }
+      expect(response).not_to be_redirect
```
Comment (DmitryGolubko) [inline spec/controllers/products_controller_spec.rb:89]: you also have to check that the new product is not created

File: spec/controllers/products_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      expect(product).to_not render_template('products/index')
+    end
+  end
+
+  describe 'POST create' do
+    let(:product) { Product.new }
+
+    it 'returns a success response' do
+      post :create, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product' } }
+      expect(response).to be_redirect
```
Comment (DmitryGolubko) [inline spec/controllers/products_controller_spec.rb:83]: you also have to check that the new product is created

File: spec/controllers/products_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  describe 'PATCH update' do
+    it 'updates the product and redirects' do
+      patch :update, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product' } }
+      expect(response).to be_redirect
+    end
+
+    it 'with bad data' do
+      patch :update, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: 1000, price: '12345RE', description: 'My Product' } }
+      expect(response).not_to be_redirect
+    end
```
Comment (DmitryGolubko) [inline spec/controllers/products_controller_spec.rb]: the same as for creation

File: spec/models/product_spec.rb [hunk 12]
```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Product, type: :model do
+  subject { Product.new(name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product') }
```
Comment (DmitryGolubko) [inline spec/models/product_spec.rb]: the same move product creating to factories

**Review by DmitryGolubko — CHANGES_REQUESTED** [review]
couple of changes

File: spec/controllers/products_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    it 'does not render a different template' do
+      expect(product).to_not render_template('products/index')
+    end
+  end
+
+  describe '#new' do
+    product = FactoryBot.create(:product)
+
+    it 'returns a success response' do
+      get :new, params: { id: product }
```
Comment (DmitryGolubko) [inline spec/controllers/products_controller_spec.rb]: Do we really need id in new template?

**Review by DmitryGolubko — CHANGES_REQUESTED** [review]
check usage of id in new template and one more comment left about creating products

**Review by DmitryGolubko — APPROVED** [review]

**Pull request merged**